std/manual.md

```diff
@@ -640,6 +646,104 @@ reference directive. If Deno used this, it would interfere with the behavior of
 the TypeScript compiler. Deno only looks for the directive in JavaScript (and
 JSX) files.
 
+### Referencing TypeScript library files
+
+When you use `deno run`, or other Deno commands which type check TypeScript,
+that code is evaluated against custom libraries which describe the environment
+that Deno supports. By default, the compiler runtime APIs which type check
+TypeScript also use these libraries (`Deno.compile()` and `Deno.bundle()`).
+
+But if you want to compile or bundle TypeScript for some other runtime, you may
+want to override the default libraries. In order to do this, the runtime APIs
+support the `lib` property in the compiler options. For example, if you had
+TypeScript code that is destined for the browser, you would want to use the
+TypeScript `"dom"` library:
+
+```ts
+const [errors, emitted] = Deno.compile(
+  "main.ts",
+  {
+    "main.ts": `document.getElementById("foo");\n`
+  },
+  {
+    lib: ["dom", "esnext"]
+  }
+);
+```
+
+For a list of all the libraries that TypeScript supports, see the
+[`lib` compiler option](https://www.typescriptlang.org/docs/handbook/compiler-options.html)
+documentation. Any libraries that aren't built into Deno, but are available in
+with TypeScript are lazily loaded and cached from the official GitHub repository
+TypeScript. This means the first time the file is required you need an active
+network connection. The file will be cached in the same way as any other
+dependency though.
```

Comment:
Hey Kitson - sorry for the slow review.

This patch is clean and well implemented, as always. I have a philosophical concern about this feature tho: I don't like how the string "dom" is implicitly turned into a download from github. Certainly the idea of downloading external code from the web is what deno does, but always those URLs are explicitly stated. With this change the user's code could not contain the string "https://github.com/microsoft/typescript", yet still result in a download from there.

I wonder if we couldn't just include lib.dom.d.ts as an asset, along with the other ones? It seems like it would add about 1mb to our executable....  which I think wouldn't be the worst thing in the world. (If we gzipped it, it would probably be much smaller)
```
-rw-r--r--  1 rld  staff   819K Jan 14 16:36 lib.dom.d.ts
-rw-r--r--  1 rld  staff   245K Jan 14 16:36 lib.webworker.d.ts
-rw-r--r--  1 rld  staff   199K Jan 14 16:36 lib.es5.d.ts
-rw-r--r--  1 rld  staff    30K Jan 14 16:36 lib.esnext.bigint.d.ts
-rw-r--r--  1 rld  staff    19K Jan 14 16:36 lib.es2015.core.d.ts
-rw-r--r--  1 rld  staff    15K Jan 14 16:36 lib.es2015.iterable.d.ts
-rw-r--r--  1 rld  staff    14K Jan 14 16:36 lib.dom.iterable.d.ts
-rw-r--r--  1 rld  staff    10K Jan 14 16:36 lib.es2015.symbol.wellknown.d.ts
-rw-r--r--  1 rld  staff   9.2K Jan 14 16:36 lib.scripthost.d.ts
-rw-r--r--  1 rld  staff   7.5K Jan 14 16:36 lib.es2019.array.d.ts
-rw-r--r--  1 rld  staff   7.5K Jan 14 16:36 lib.esnext.array.d.ts
-rw-r--r--  1 rld  staff   7.1K Jan 14 16:36 lib.es2015.promise.d.ts
-rw-r--r--  1 rld  staff   6.0K Jan 14 16:36 lib.es2017.sharedmemory.d.ts
-rw-r--r--  1 rld  staff   4.8K Jan 14 16:36 lib.es2016.array.include.d.ts
-rw-r--r--  1 rld  staff   2.8K Jan 14 16:36 lib.es2015.collection.d.ts
-rw-r--r--  1 rld  staff   2.6K Jan 14 16:36 lib.es2018.asyncgenerator.d.ts
-rw-r--r--  1 rld  staff   2.5K Jan 14 16:36 lib.es2015.generator.d.ts
-rw-r--r--  1 rld  staff   2.4K Jan 14 16:36 lib.es2017.object.d.ts
-rw-r--r--  1 rld  staff   2.3K Jan 14 16:36 lib.es2017.string.d.ts
-rw-r--r--  1 rld  staff   1.9K Jan 14 16:36 lib.es2015.proxy.d.ts
-rw-r--r--  1 rld  staff   1.9K Jan 14 16:36 lib.es2015.reflect.d.ts
-rw-r--r--  1 rld  staff   1.8K Jan 14 16:36 lib.es2018.intl.d.ts
-rw-r--r--  1 rld  staff   1.7K Jan 14 16:36 lib.es2018.asynciterable.d.ts
-rw-r--r--  1 rld  staff   1.6K Jan 14 16:36 lib.es2015.symbol.d.ts
-rw-r--r--  1 rld  staff   1.5K Jan 14 16:36 lib.esnext.asynciterable.d.ts
-rw-r--r--  1 rld  staff   1.5K Jan 14 16:36 lib.es2019.object.d.ts
-rw-r--r--  1 rld  staff   1.4K Jan 14 16:36 lib.es2020.symbol.wellknown.d.ts
-rw-r--r--  1 rld  staff   1.4K Jan 14 16:36 lib.es2017.typedarrays.d.ts
-rw-r--r--  1 rld  staff   1.3K Jan 14 16:36 lib.es2018.promise.d.ts
-rw-r--r--  1 rld  staff   1.3K Jan 14 16:36 lib.es2019.string.d.ts
-rw-r--r--  1 rld  staff   1.2K Jan 14 16:36 lib.es2017.intl.d.ts
-rw-r--r--  1 rld  staff   1.2K Jan 14 16:36 lib.esnext.intl.d.ts
-rw-r--r--  1 rld  staff   1.2K Jan 14 16:36 lib.es2015.d.ts
-rw-r--r--  1 rld  staff   1.2K Jan 14 16:36 lib.es2020.string.d.ts
-rw-r--r--  1 rld  staff   1.2K Jan 14 16:36 lib.es2018.regexp.d.ts
-rw-r--r--  1 rld  staff   1.1K Jan 14 16:36 lib.es2018.d.ts
-rw-r--r--  1 rld  staff   1.1K Jan 14 16:36 lib.es2017.d.ts
-rw-r--r--  1 rld  staff   1.0K Jan 14 16:36 lib.webworker.importscripts.d.ts
-rw-r--r--  1 rld  staff   1.0K Jan 14 16:36 lib.es2019.d.ts
-rw-r--r--  1 rld  staff   1.0K Jan 14 16:36 lib.es2019.full.d.ts
-rw-r--r--  1 rld  staff   1.0K Jan 14 16:36 lib.es2020.full.d.ts
-rw-r--r--  1 rld  staff   1.0K Jan 14 16:36 lib.es6.d.ts
-rw-r--r--  1 rld  staff   1.0K Jan 14 16:36 lib.es2016.full.d.ts
-rw-r--r--  1 rld  staff   1.0K Jan 14 16:36 lib.es2017.full.d.ts
-rw-r--r--  1 rld  staff   1.0K Jan 14 16:36 lib.es2018.full.d.ts
-rw-r--r--  1 rld  staff   1.0K Jan 14 16:36 lib.esnext.full.d.ts
-rw-r--r--  1 rld  staff   1.0K Jan 14 16:36 lib.es2019.symbol.d.ts
-rw-r--r--  1 rld  staff   1.0K Jan 14 16:36 lib.esnext.symbol.d.ts
-rw-r--r--  1 rld  staff   1.0K Jan 14 16:36 lib.d.ts
-rw-r--r--  1 rld  staff   979B Jan 14 16:36 lib.es2020.d.ts
-rw-r--r--  1 rld  staff   967B Jan 14 16:36 lib.esnext.d.ts
-rw-r--r--  1 rld  staff   937B Jan 14 16:36 lib.es2016.d.ts
```

Alternatively I would be happy to require users to supply a URL completely

```ts
/// <reference lib="https://raw.githubusercontent.com/microsoft/TypeScript/v3.7.2/lib/lib.dom.d.ts"/>
```

Reply:
PR builds them in now.